Source/JavaScriptCore/runtime/VMTraps.h

```diff
@@ -244,9 +244,15 @@ class VMTraps {
     ALWAYS_INLINE CONCURRENT_SAFE void fireTrap(Event event)
     {
         ASSERT(!(event & ~AllEvents));
+
+        if (isAsyncEvent(event)) {
+            Locker locker { *m_trapSignalingLock };
+            m_trapBits.exchangeOr(event);
+            requestThreadStopIfNeeded(locker, event);
+            return;
+        }
+
         m_trapBits.exchangeOr(event);
-        if (isAsyncEvent(event))
-            requestThreadStopIfNeeded(event);
```

Comment:
This is incorrect.  m_trapBits must always be set before we request the stop.  Otherwise, a race can occur where the stop occurs but canot see the requested trap bit yet.

I think the right fix is to remove the `RELEASE_ASSERT(m_threadStopRequested);` in `VMTraps::cancelThreadStopIfNeeded()` since it is obvious that we cannot guarantee that `m_threadStopRequested` is set by the time we get there.  There's already a check after that that returns early if already `m_threadStopRequested`.  Otherwise, it will request the thread stop.

Reply:
Discussed offline with Mark and Dan. We should introduces `updateThreadStopRequestIfNeeded()`, which synchronizes `m_threadStopRequested` to match `m_trapBits` regardless of the current state.